File: libs/power-log-parser/src/lib/models/node.ts
export enum NodeType {
	Action = 'Action',
	ChangeEntity = 'ChangeEntity',
	Choice = 'Choice',
	Choices = 'Choices',
	FullEntity = 'FullEntity',
	Game = 'Game',
	GameAction = 'GameAction',
	GameEntity = 'GameEntity',
	HideEntity = 'HideEntity',
	Info = 'Info',
	MetaData = 'MetaData',
	Placeholder = 'Placeholder',
	PlayerEntity = 'PlayerEntity',
	ShowEntity = 'ShowEntity',
	ShuffleDeck = 'ShuffleDeck',
	SubSpell = 'SubSpell',
	TagChange = 'TagChange',
}

export class Node {
	private static currentIndex: number = 0;

	Type: NodeType;
	Object: any;
	IndentLevel: number;
	Parent: Node | null;
	CreationLogLine: string;
	Index: number;
	Closed: boolean = false;

	constructor(type: NodeType, o: any, indentLevel: number, parent: Node | null, creationLogLine: string) {
		this.Type = type;
		this.Object = o;
		this.IndentLevel = indentLevel;
		this.Parent = parent;
		this.CreationLogLine = creationLogLine;
		if (creationLogLine == null) {
			throw new Error(`Should not create nodes with empty creationLogLine: ${type}`);
		}
		this.Index = Node.currentIndex++;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/Node.cs
#region

using System;

#endregion

namespace HearthstoneReplays.Parser
{
	public class Node
	{
		private static int currentIndex = 0;

		public Node(Type type, object o, int indentLevel, Node parent, string creationLogLine)
		{
			Type = type;
			Object = o;
			IndentLevel = indentLevel;
			Parent = parent;
            CreationLogLine = creationLogLine;
            if (creationLogLine == null)
            {
                throw new Exception("Should not create nodes with empty creationLogLine: "
                    + type.ToString());
            }
			Index = currentIndex++;
		}

		public Type Type { get; set; }
		public object Object { get; set; }
		public int IndentLevel { get; set; }
		public Node Parent { get; set; }
		public string CreationLogLine { get; set; }
		public int Index { get; set; }
		public bool Closed { get; set; }
	}
}